sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -104,8 +111,40 @@ private object DateTimeFormatterHelper {
       .withResolverStyle(ResolverStyle.STRICT)
   }
 
-  def buildFormatter(pattern: String, locale: Locale): DateTimeFormatter = {
-    val builder = createBuilder().appendPattern(pattern)
+  /**
+   * Building a formatter for parsing seconds fraction with variable length
+   */
+  def appendPattern(
+      pattern: String,
+      builder: DateTimeFormatterBuilder): DateTimeFormatterBuilder = {
+    pattern.split("'").zipWithIndex.foreach {
+      case (pattenPart, idx) if idx % 2 == 0 =>
+        var rest = pattenPart
+        while (rest.nonEmpty) {
+          rest match {
+            case extractor(prefix, sss, suffix) =>
+              builder.appendPattern(prefix)
+              if (sss.nonEmpty) {
+                builder.appendFraction(ChronoField.NANO_OF_SECOND, 0, sss.length, false)
+              }
+              rest = suffix
+            case _ => // never reach
```

Comment:
it's reachable. The outer-most is `pattern.split("'")`, so it's possible that a section of pattern string doesn't have `S`.

Reply:
the `extractor` here can match anything, so this is unreachable. If a pattern split contains no `S`,  it goes the `prefix`

Reply:
then let's throw `IllegalStateException` as it's un-reachable